webrisk/webrisk_client.go

```diff
@@ -0,0 +1,522 @@
+				for _, td := range unsureThreats {
+					if _, ok := cachedThreats[td]; ok {
+						threats[i] = append(threats[i], URLThreat{
+							Pattern:          pattern,
+							ThreatType: td,
+						})
+					}
+				}
+				atomic.AddInt64(&sb.stats.QueriesByCache, 1)
+			case negativeCacheHit:
+				// This is cached as a non-threat.
+				atomic.AddInt64(&sb.stats.QueriesByCache, 1)
+				continue
+			default:
+				// The cache knows nothing about this full hash, so we must make
+				// a request for it.
+				if alreadyRequested {
+					continue
+				}
+				for _, td := range unsureThreats {
+					ttm[pb.ThreatType(td)] = true
+				}
+
+				tts := []pb.ThreatType{}
+				for _, tt := range unsureThreats {
+					tts = append(tts, pb.ThreatType(tt))
+				}
+
+				reqs = append(reqs, &pb.SearchHashesRequest {
+					HashPrefix: []byte(partialHash),
+					ThreatTypes: tts,
+				})
+			}
+		}
+	}
+
+	for _, req := range reqs {
+		// Actually query the Web Risk API for exact full hash matches.
+		resp, err := sb.api.HashLookup(ctx, req)
+		if err != nil {
+			sb.log.Printf("HashLookup failure: %v", err)
+			atomic.AddInt64(&sb.stats.QueriesFail, 1)
+			return threats, err
+		}
+
+		// Update the cache.
+		sb.c.Update(req, resp)
+
+		// Pull the information the client cares about out of the response.
+		for _, threat := range resp.GetThreats() {
+			fullHash := hashPrefix(threat.Hash)
+			if !fullHash.IsFull() {
+				continue
+			}
+			pattern, ok := hashes[fullHash]
+			idxs, findidx := hash2idxs[fullHash]
+			if findidx && ok {
+				for _, td := range threat.ThreatTypes {
+					if !sb.lists[ThreatType(td)] {
+						continue
+					}
+					for _, idx := range idxs {
+						threats[idx] = append(threats[idx], URLThreat{
+							Pattern:          pattern,
+							ThreatType: ThreatType(td),
+						})
+					}
+				}
+			}
+		}
+		atomic.AddInt64(&sb.stats.QueriesByAPI, 1)
+	}
+	return threats, nil
+}
+
+// TODO: Add other types of lookup when available.
+//	func (sb *WebriskClient) LookupBinaries(digests []string) (threats []BinaryThreat, err error)
+//	func (sb *WebriskClient) LookupAddresses(addrs []string) (threats [][]AddressThreat, err error)
+
+// updater is a blocking method that periodically updates the local database.
+// This should be run as a separate goroutine and will be automatically stopped
+// when sb.Close is called.
+func (sb *WebriskClient) updater(delay time.Duration) {
+	for {
+		sb.log.Printf("Next update in %v", delay)
+		select {
+		case <-time.After(delay):
+			var ok bool
+			ctx, cancel := context.WithTimeout(context.Background(), sb.config.RequestTimeout)
+			if delay, ok = sb.db.Update(ctx, sb.api); ok {
+				sb.log.Printf("background threat list updated")
+				sb.c.Purge()
+			}
+			cancel()
+
+		case <-sb.done:
+			return
+		}
+	}
+}
+
+// Close cleans up all resources.
+// This method must not be called concurrently with other lookup methods.
+func (sb *WebriskClient) Close() error {
+	if atomic.LoadUint32(&sb.closed) == 0 {
```

Comment:
Not sure of a better way to do it, but this seems weird.

Reply:
ack.